INF 510, Lab #5
---------------

Thursday, February 7th 2019 (4:30 - 6:20pm)



### Lab rules

1.  Please read the instructions and problem prompts **carefully**.  Be sure to solve the problem the way it is asked (i.e. are you supposed to define a function?  Catch exceptions?  Use a particular construct?)  Solutions that "work", but are not composed according to the directions will not receive [full] credit

2.  This lab is to be done *individually*. You may talk to your fellow classmates about general issues (*"Remind me again: What's the difference between a parameter and an argument?"*) but not about the specifics of how to do these exercises.

3.  Along a similar vein, you can ask the TA for help, but ask questions about **concepts**.  (i.e. *"I don’t understand how the else block works"*, not "*my program isn’t doing what I want"*).  The TA is there to help, but not to do your work for you.

4.  You are welcome to use the class resources and the Internet. But try to google for *concepts*, not your specific tasks.  The goal for you to begin to start thinking algorithmically, and hopefully memorize a little syntax by osmosis. If you just blindly do the exercises, without actually *thinking* about them, the next time you see something related but slightly altered — like, say, on an exam — you won’t be able to solve the problem. So avoid brainlessly cutting and pasting code snippets.

5.  Play with variations. Solve one of the problems, and then copy the code to a new cell and play around with it. Experiment! Break stuff! Doing this is **the single most important thing** when learning a programming language.

6.  When you have finished all the exercises, let the TA know, and he will check your work.  Presuming you’ve done things correctly, you’re free to go.

7.  Have fun!



### Problem 1 (1 points)

Think about a program which opens a .CSV file, reads in the "header row", uses the elements of that row to create dictionaries with the elements in the header row as dictionary keys, and then reads each line in the rest of the .CSV file as dictionary values that correspond to those keys.  Keep in mind, this should be general; it shouldn't matter what the CSV looks like, other than the first row is header information and the rest are data.  You do not need to catch exceptions here.

How might you design this program?  Think about the different "parts" it contains: You have to open the file, read the contents, do some special handling of the header, create the keys, create a dictionary entry for each line in the CSV etc. 

Write down in English -- or in *pseudo-code* -- how this program might look.  For example, if I was going to write a program that added up every number in a list of numbers (checking to make sure the elements of a list were numbers) I might write the following:

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
set some running total to zero
iterate through the list
for each element in the list, I need to do some stuff:
     check if the element is a legit number
     if it is then I should
          add the number to my running total
     if it is not legit then I should
          tell the user to do it right next time and move on to the next element
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Write your answer in the next cell, replacing the "Replace this with your answer" text in between the dashed lines.  Be **sure** to leave the "-----------" lines there. 





~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
handle = open('the name of csv file')\
set a dict for output
set a list for dict key
set a list for dict value
for index,content in enumerate(output_dic_key):
    temp = []
    output_dic_value.append(temp)
for line in handle:
    temp = line.split(',')
    for index, content in enumerate(temp):
        output_dic_value[index].append(content) 
for index, content in enumerate(output_dic_key):
    output_dic[content] = output_dic_value[index]


~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~




### Problem 2 (6 points)

Actually implmement the program described in Problem 1.  I've included some sample code on  the **zip** function.  Explore a bit what the function does, and take a look at the documentation here (https://docs.python.org/3/library/functions.html#zip).

You can use the file from last week's lab for sample data (although any well-formed CSV should work)

In [1]:
int_list = [1,3,3,4]
string_list = ['one', 'two', 'three', 'four']
zipped = zip(int_list, string_list)
zipped_list = list(zipped)  # zip returns an iterator, so lets convert it to a list
print(zipped_list)

[(1, 'one'), (3, 'two'), (3, 'three'), (4, 'four')]


Note what **zip** does: It iterates across two *iteratable* objects, pairing the first element of each into a tuple, the second element into another tuple, and so on.  Given that we're pairing headers (*keys*) with data rows (*values*) this might be handy. =]

In [24]:
handle = open('week_1_probabilities.csv')
output_dic = {}
output_dic_key = list(handle.readline().split(','))
output_dic_value = []

for index,content in enumerate(output_dic_key):
    temp = []
    output_dic_value.append(temp)
for line in handle:
    temp = line.split(',')
    for index, content in enumerate(temp):
        output_dic_value[index].append(content) 
for index, content in enumerate(output_dic_key):
    output_dic[content] = output_dic_value[index]
        
print((output_dic_key))    
print(len(output_dic_value))
print(output_dic)

['home team city', 'home team name', 'home team probability', 'away team city', 'away team name', 'away team probability', 'tie probabailty\n']
7
{'home team city': ['New England', 'Buffalo', 'Chicago', 'Cincinnati', 'Cleveland', 'Detroit', 'Houston', 'Tennessee', 'Washington', 'Los Angeles', 'Green Bay', 'San Francisco', 'Dallas', 'Minnesota', 'Denver'], 'home team name': ['Patriots', 'Bills', 'Bears', 'Bengals', 'Browns', 'Lions', 'Texans', 'Titans', 'Redskins', 'Rams', 'Packers', '49ers', 'Cowboys', 'Vikings', 'Broncos'], 'home team probability': ['74.6', '75.2', '35.9', '58.5', '25.9', '59.1', '70.2', '53.6', '53.1', '64.3', '62.0', '37.4', '66.6', '64.1', '61.5'], 'away team city': ['Kansas City', 'New York', 'Atlanta', 'Baltimore', 'Pittsburgh', 'Arizona', 'Jacksonville', 'Oakland', 'Philadelphia', 'Indianapolis', 'Seattle', 'Carolina', 'New York', 'New Orleans', 'Los Angeles'], 'away team name': ['Chiefs', 'Jets', 'Falcons', 'Ravens', 'Steelers', 'Cardinals', 'Jaguars', 'Raiders

### Problem 3 (1 point)
Imagine you want to create a new key for the dictionary[ies] above called "winner", and add it to each dictionary.  The value of the key will be whichever of the probabilities (away or home) is greater. Use the "team name" field for this value. **NOTE: If the probabilities of both teams are equal, there is no clear winner and "toss-up" should be reported"** 


i.e. before, an example entry in your dict might be:
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
{'away team city': 'Kansas City',
 'away team name': 'Chiefs',
 'away team probability': '25.2',
 'home team city': 'New England',
 'home team name': 'Patriots',
 'home team probability': '74.6',
 'tie probabailty': '2.0'}
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


 but after, you'll have:
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
{'away team city': 'Kansas City',
 'away team name': 'Chiefs',
 'away team probability': '25.2',
 'home team city': 'New England',
 'home team name': 'Patriots',
 'home team probability': '74.6',
 'tie probabailty': '2.0'
 'winner': 'Patriots'}
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Note that this is because the Patriots have a greater probabiltiy than the Chiefs (74.6 > 25.2).  So we use the team name of "Patriots".

Then, iterate through your dictionaries that you created above, adding this new field (and new values).  Then write these new dictionaries back to a CSV file called lab5_output.csv.  You'll want to use the keys as a header, then write the data as rows, comma-separated.  Feel free to use code from Lab 4 if you'd like, as it might be similar.

Write psuedo-code for how to do this in the cell below.  Think about the steps you'll need to do, and what loops, conditionals, data structures, etc. you'll need.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Replace this text with your answer to Problem 3.  Leave the dashed lines above and below.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~





### Problem 4 (3 points)
Implement what you sketched out above.
 

In [63]:
handle = open('week_1_probabilities.csv')
output_dic = {}
output_dic_key = list(handle.readline().split(','))
output_dic_key.append('winner')
output_dic_value = []

for index,content in enumerate(output_dic_key):
    temp = []
    output_dic_value.append(temp)
for line in handle:
    temp = line.split(',')
    for index, content in enumerate(temp):
        output_dic_value[index].append(content)
    output_dic_value[len(output_dic_key)-1].append(0)
for index, content in enumerate(output_dic_key):
    output_dic[content] = output_dic_value[index]
     

        
for index, content in enumerate(output_dic['winner']):
    if output_dic['away team probability'][index] > output_dic['home team probability'][index]:
        output_dic['winner'][index] = (output_dic['away team name'][index])
    elif output_dic['away team probability'][index] < output_dic['home team probability'][index]:
        output_dic['winner'][index] = (output_dic['home team name'][index])
    else:
         output_dic['winner'][index] = 'hard to decide'
        
        

        
print(output_dic)          

{'home team city': ['New England', 'Buffalo', 'Chicago', 'Cincinnati', 'Cleveland', 'Detroit', 'Houston', 'Tennessee', 'Washington', 'Los Angeles', 'Green Bay', 'San Francisco', 'Dallas', 'Minnesota', 'Denver'], 'home team name': ['Patriots', 'Bills', 'Bears', 'Bengals', 'Browns', 'Lions', 'Texans', 'Titans', 'Redskins', 'Rams', 'Packers', '49ers', 'Cowboys', 'Vikings', 'Broncos'], 'home team probability': ['74.6', '75.2', '35.9', '58.5', '25.9', '59.1', '70.2', '53.6', '53.1', '64.3', '62.0', '37.4', '66.6', '64.1', '61.5'], 'away team city': ['Kansas City', 'New York', 'Atlanta', 'Baltimore', 'Pittsburgh', 'Arizona', 'Jacksonville', 'Oakland', 'Philadelphia', 'Indianapolis', 'Seattle', 'Carolina', 'New York', 'New Orleans', 'Los Angeles'], 'away team name': ['Chiefs', 'Jets', 'Falcons', 'Ravens', 'Steelers', 'Cardinals', 'Jaguars', 'Raiders', 'Eagles', 'Colts', 'Seahawks', 'Panthers', 'Giants', 'Saints', 'Chargers'], 'away team probability': ['25.2', '24.6', '63.8', '41.2', '73.9', '

In [83]:
output_str = ''
for index, content in enumerate(output_dic_key):
    output_str = output_str + ',' + output_dic_key[index] + ','



In [84]:
f = open('lab5_ouput.csv','w')
f.write(output_str)
f.close()

In [31]:
for i in a:
    print (i)

a
b
c


In [46]:
a = [1,2,3]

In [47]:
len(a)

3